x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDatafeedsStatsAction.java

```diff
@@ -58,22 +63,37 @@ protected void masterOperation(GetDatafeedsStatsAction.Request request, ClusterS
                                    ActionListener<GetDatafeedsStatsAction.Response> listener) throws Exception {
         logger.debug("Get stats for datafeed '{}'", request.getDatafeedId());
 
-        datafeedConfigProvider.expandDatafeedIds(request.getDatafeedId(), request.allowNoDatafeeds(), ActionListener.wrap(
-                expandedDatafeedIds -> {
-                    PersistentTasksCustomMetaData tasksInProgress = state.getMetaData().custom(PersistentTasksCustomMetaData.TYPE);
-                    List<GetDatafeedsStatsAction.Response.DatafeedStats> results = expandedDatafeedIds.stream()
-                            .map(datafeedId -> getDatafeedStats(datafeedId, state, tasksInProgress))
-                            .collect(Collectors.toList());
-                    QueryPage<GetDatafeedsStatsAction.Response.DatafeedStats> statsPage = new QueryPage<>(results, results.size(),
-                            DatafeedConfig.RESULTS_FIELD);
-                    listener.onResponse(new GetDatafeedsStatsAction.Response(statsPage));
+        datafeedConfigProvider.expandDatafeedConfigs(
+            request.getDatafeedId(),
+            request.allowNoDatafeeds(),
+            ActionListener.wrap(
+                datafeedBuilders -> {
+                    Map<String, String> jobIdByDatafeedId = datafeedBuilders.stream()
+                        .map(DatafeedConfig.Builder::build)
+                        .collect(Collectors.toUnmodifiableMap(DatafeedConfig::getId, DatafeedConfig::getJobId));
```

Comment:
It seems to me that this whole piece of logic is required because you are not also storing the `datafeed_id` with the datafeed stats document. It seems much simpler to me to save the datafeed_id so you don't have to make multiple es client calls to read docs.

Reply:
I could store datafeed_id in the timing_stats doc, but:
1. the number of calls to es client here would remain unchanged (one to fetch expanded datafeed ids, another one to fetch timing stats for those datafeed ids)
2. in JobResultsProvider::datafeedTimingStats, I couldn't use the IdsQuery anymore (for ids query I need job ids in order to transform them into document ids).
So I do not see much gain from adding datafeed_id.
Happy to chat priv if I missed something or didn't make my point clear enough.

Reply:
@przemekwitek You are correct :)